### ISSUE TITLE
Configuring Exchange fails

### ISSUE THREAD

**Issue by Oleszkiewicz** [zammad/zammad#2065]
### Infos:

* Used Zammad version: 2.4.x (newest installable)
* Installation method (source, package, ..): docker-compose
* Operating system: Photon OS 2.0
* Database + version: postgres:9.6-alpine
* Elasticsearch version: 5.6.9
* Browser + version: any

### Expected behavior:

Zammad connects to Exchange

### Actual behavior:

Cannot assign requested address - Cannot assign requested address - connect(2) for "autodiscover.xxxxxx.com" port 443 (autodiscover.xxxxx.com:443)

### Steps to reproduce the behavior:

Put in the username and password, click connect

Yes I'm sure this is a bug and no feature request or a general question.

**Comment by muhammadn** [comment]
Hi @Oleszkiewicz 

This is not a bug but a technical question. For technical questions, please go to https://community.zammad.org

Thank you.

**Comment by thorsteneckel** [comment]
This actually might be a bug. @Oleszkiewicz could you please provide the `log/production.log` file that shows the error message and the stack trace showing where the error occurred? Thanks!

**Comment by Oleszkiewicz** [comment]
```
zammad-nginx_1          | 192.168.2.1 - - [11/Jun/2018:23:48:42 +0000] "GET /api/v1/integration/exchange/job_start?_=1528760852423 HTTP/1.1" 200 33 "http://10.110.10.15/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:60.0) Gecko/20100101 Firefox/60.0"
zammad-railsserver_1    | I, [2018-06-11T23:48:43.427495 #1]  INFO -- : Started POST "/api/v1/integration/exchange/autodiscover" for ::ffff:172.18.0.10 at 2018-06-11 23:48:43 +0000
zammad-railsserver_1    | I, [2018-06-11T23:48:43.434155 #1]  INFO -- : Processing by Integration::ExchangeController#autodiscover as JSON
zammad-railsserver_1    | I, [2018-06-11T23:48:43.434219 #1]  INFO -- :   Parameters: {"user"=>"piotr.oleszkiewicz@xxxxxxx.com", "password"=>"[FILTERED]"}
zammad-railsserver_1    | E, [2018-06-11T23:48:43.710286 #1] ERROR -- : Cannot assign requested address - Cannot assign requested address - connect(2) for "autodiscover.xxxxxxx.com" port 443 (autodiscover.xxxxxxx.com:443) (Errno::EADDRNOTAVAIL)
zammad-railsserver_1    | /usr/local/bundle/gems/httpclient-2.8.3/lib/httpclient/session.rb:623:in `rescue in create_socket'
zammad-railsserver_1    | /usr/local/bundle/gems/httpclient-2.8.3/lib/httpclient/session.rb:607:in `create_socket'
zammad-railsserver_1    | /usr/local/bundle/gems/httpclient-2.8.3/lib/httpclient/ssl_socket.rb:21:in `create_socket'
zammad-railsserver_1    | /usr/local/bundle/gems/httpclient-2.8.3/lib/httpclient/session.rb:752:in `block in connect'
zammad-railsserver_1    | /usr/local/lib/ruby/2.4.0/timeout.rb:93:in `block in timeout'
zammad-railsserver_1    | /usr/local/lib/ruby/2.4.0/timeout.rb:103:in `timeout'
zammad-railsserver_1    | /usr/local/bundle/gems/httpclient-2.8.3/lib/httpclient/session.rb:748:in `connect'
zammad-railsserver_1    | /usr/local/bundle/gems/httpclient-2.8.3/lib/httpclient/session.rb:511:in `query'
zammad-railsserver_1    | /usr/local/bundle/gems/httpclient-2.8.3/lib/httpclient/session.rb:177:in `query'
zammad-railsserver_1    | /usr/local/bundle/gems/httpclient-2.8.3/lib/httpclient.rb:1242:in `do_get_block'
zammad-railsserver_1    | /usr/local/bundle/gems/httpclient-2.8.3/lib/httpclient.rb:1019:in `block in do_request'
zammad-railsserver_1    | /usr/local/bundle/gems/httpclient-2.8.3/lib/httpclient.rb:1133:in `protect_keep_alive_disconnected'
zammad-railsserver_1    | /usr/local/bundle/gems/httpclient-2.8.3/lib/httpclient.rb:1014:in `do_request'
zammad-railsserver_1    | /usr/local/bundle/gems/httpclient-2.8.3/lib/httpclient.rb:856:in `request'
zammad-railsserver_1    | /usr/local/bundle/gems/httpclient-2.8.3/lib/httpclient.rb:765:in `post'
zammad-railsserver_1    | /usr/local/bundle/bundler/gems/autodiscover-ee9b53dfa797/lib/autodiscover/pox_request.rb:19:in `block in autodiscover'
zammad-railsserver_1    | /usr/local/bundle/bundler/gems/autodiscover-ee9b53dfa797/lib/autodiscover/pox_request.rb:35:in `block (2 levels) in available_urls'
zammad-railsserver_1    | /usr/local/bundle/bundler/gems/autodiscover-ee9b53dfa797/lib/autodiscover/pox_request.rb:71:in `handle_allowed_errors'
zammad-railsserver_1    | /usr/local/bundle/bundler/gems/autodiscover-ee9b53dfa797/lib/autodiscover/pox_request.rb:34:in `block in available_urls'
zammad-railsserver_1    | /usr/local/bundle/bundler/gems/autodiscover-ee9b53dfa797/lib/autodiscover/pox_request.rb:32:in `each'
zammad-railsserver_1    | /usr/local/bundle/bundler/gems/autodiscover-ee9b53dfa797/lib/autodiscover/pox_request.rb:32:in `available_urls'
zammad-railsserver_1    | /usr/local/bundle/bundler/gems/autodiscover-ee9b53dfa797/lib/autodiscover/pox_request.rb:18:in `each'
zammad-railsserver_1    | /usr/local/bundle/bundler/gems/autodiscover-ee9b53dfa797/lib/autodiscover/pox_request.rb:18:in `autodiscover'
zammad-railsserver_1    | /usr/local/bundle/bundler/gems/autodiscover-ee9b53dfa797/lib/autodiscover/client.rb:27:in `autodiscover'
zammad-railsserver_1    | /usr/local/bundle/gems/activesupport-5.1.4/lib/active_support/core_ext/object/try.rb:17:in `public_send'
zammad-railsserver_1    | /usr/local/bundle/gems/activesupport-5.1.4/lib/active_support/core_ext/object/try.rb:17:in `try!'
zammad-railsserver_1    | /usr/local/bundle/gems/activesupport-5.1.4/lib/active_support/core_ext/object/try.rb:6:in `try'
zammad-railsserver_1    | /opt/zammad/app/controllers/integration/exchange_controller.rb:20:in `block in autodiscover'
zammad-railsserver_1    | /opt/zammad/app/controllers/concerns/integration/import_job_base.rb:43:in `answer_with'
zammad-railsserver_1    | /opt/zammad/app/controllers/integration/exchange_controller.rb:9:in `autodiscover'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/action_controller/metal/basic_implicit_render.rb:4:in `send_action'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/abstract_controller/base.rb:186:in `process_action'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/action_controller/metal/rendering.rb:30:in `process_action'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/abstract_controller/callbacks.rb:20:in `block in process_action'
zammad-railsserver_1    | /usr/local/bundle/gems/activesupport-5.1.4/lib/active_support/callbacks.rb:131:in `run_callbacks'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/abstract_controller/callbacks.rb:19:in `process_action'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/action_controller/metal/rescue.rb:20:in `process_action'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/action_controller/metal/instrumentation.rb:32:in `block in process_action'
zammad-railsserver_1    | /usr/local/bundle/gems/activesupport-5.1.4/lib/active_support/notifications.rb:166:in `block in instrument'
zammad-railsserver_1    | /usr/local/bundle/gems/activesupport-5.1.4/lib/active_support/notifications/instrumenter.rb:21:in `instrument'
zammad-railsserver_1    | /usr/local/bundle/gems/activesupport-5.1.4/lib/active_support/notifications.rb:166:in `instrument'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/action_controller/metal/instrumentation.rb:30:in `process_action'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/action_controller/metal/params_wrapper.rb:252:in `process_action'
zammad-railsserver_1    | /usr/local/bundle/gems/activerecord-5.1.4/lib/active_record/railties/controller_runtime.rb:22:in `process_action'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/abstract_controller/base.rb:124:in `process'
zammad-railsserver_1    | /usr/local/bundle/gems/actionview-5.1.4/lib/action_view/rendering.rb:30:in `process'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/action_controller/metal.rb:189:in `dispatch'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/action_controller/metal.rb:253:in `dispatch'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/action_dispatch/routing/route_set.rb:49:in `dispatch'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/action_dispatch/routing/route_set.rb:31:in `serve'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/action_dispatch/journey/router.rb:50:in `block in serve'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/action_dispatch/journey/router.rb:33:in `each'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/action_dispatch/journey/router.rb:33:in `serve'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/action_dispatch/routing/route_set.rb:834:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/omniauth-1.7.1/lib/omniauth/strategy.rb:190:in `call!'
zammad-railsserver_1    | /usr/local/bundle/gems/omniauth-1.7.1/lib/omniauth/strategy.rb:168:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/omniauth-1.7.1/lib/omniauth/strategy.rb:190:in `call!'
zammad-railsserver_1    | /usr/local/bundle/gems/omniauth-1.7.1/lib/omniauth/strategy.rb:168:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/omniauth-1.7.1/lib/omniauth/strategy.rb:190:in `call!'
zammad-railsserver_1    | /usr/local/bundle/gems/omniauth-1.7.1/lib/omniauth/strategy.rb:168:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/omniauth-1.7.1/lib/omniauth/strategy.rb:190:in `call!'
zammad-railsserver_1    | /usr/local/bundle/gems/omniauth-1.7.1/lib/omniauth/strategy.rb:168:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/omniauth-1.7.1/lib/omniauth/strategy.rb:190:in `call!'
zammad-railsserver_1    | /usr/local/bundle/gems/omniauth-1.7.1/lib/omniauth/strategy.rb:168:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/omniauth-1.7.1/lib/omniauth/strategy.rb:190:in `call!'
zammad-railsserver_1    | /usr/local/bundle/gems/omniauth-1.7.1/lib/omniauth/strategy.rb:168:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/omniauth-1.7.1/lib/omniauth/strategy.rb:190:in `call!'
zammad-railsserver_1    | /usr/local/bundle/gems/omniauth-1.7.1/lib/omniauth/strategy.rb:168:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/omniauth-1.7.1/lib/omniauth/strategy.rb:190:in `call!'
zammad-railsserver_1    | /usr/local/bundle/gems/omniauth-1.7.1/lib/omniauth/strategy.rb:168:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/omniauth-1.7.1/lib/omniauth/strategy.rb:190:in `call!'
zammad-railsserver_1    | /usr/local/bundle/gems/omniauth-1.7.1/lib/omniauth/strategy.rb:168:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/omniauth-1.7.1/lib/omniauth/builder.rb:63:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/rack-2.0.3/lib/rack/etag.rb:25:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/rack-2.0.3/lib/rack/conditional_get.rb:38:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/rack-2.0.3/lib/rack/head.rb:12:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/rack-2.0.3/lib/rack/session/abstract/id.rb:232:in `context'
zammad-railsserver_1    | /usr/local/bundle/gems/rack-2.0.3/lib/rack/session/abstract/id.rb:226:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/action_dispatch/middleware/cookies.rb:613:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/action_dispatch/middleware/callbacks.rb:26:in `block in call'
zammad-railsserver_1    | /usr/local/bundle/gems/activesupport-5.1.4/lib/active_support/callbacks.rb:97:in `run_callbacks'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/action_dispatch/middleware/callbacks.rb:24:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/action_dispatch/middleware/debug_exceptions.rb:59:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/action_dispatch/middleware/show_exceptions.rb:31:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/railties-5.1.4/lib/rails/rack/logger.rb:36:in `call_app'
zammad-railsserver_1    | /usr/local/bundle/gems/railties-5.1.4/lib/rails/rack/logger.rb:26:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/action_dispatch/middleware/remote_ip.rb:79:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/action_dispatch/middleware/request_id.rb:25:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/rack-2.0.3/lib/rack/method_override.rb:22:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/rack-2.0.3/lib/rack/runtime.rb:22:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/activesupport-5.1.4/lib/active_support/cache/strategy/local_cache_middleware.rb:27:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/actionpack-5.1.4/lib/action_dispatch/middleware/executor.rb:12:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/rack-2.0.3/lib/rack/sendfile.rb:111:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/railties-5.1.4/lib/rails/engine.rb:522:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/puma-3.11.0/lib/puma/configuration.rb:225:in `call'
zammad-railsserver_1    | /usr/local/bundle/gems/puma-3.11.0/lib/puma/server.rb:624:in `handle_request'
zammad-railsserver_1    | /usr/local/bundle/gems/puma-3.11.0/lib/puma/server.rb:438:in `process_client'
zammad-railsserver_1    | /usr/local/bundle/gems/puma-3.11.0/lib/puma/server.rb:302:in `block in run'
zammad-railsserver_1    | /usr/local/bundle/gems/puma-3.11.0/lib/puma/thread_pool.rb:120:in `block in spawn_thread'
zammad-railsserver_1    | /usr/local/bundle/gems/logging-2.2.2/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'
zammad-railsserver_1    | I, [2018-06-11T23:48:43.715150 #1]  INFO -- : Completed 200 OK in 281ms (Views: 0.2ms | ActiveRecord: 2.5ms)
zammad-nginx_1          | 192.168.2.1 - - [11/Jun/2018:23:48:43 +0000] "POST /api/v1/integration/exchange/autodiscover HTTP/1.1" 200 143 "http://10.110.10.15/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:60.0) Gecko/20100101 Firefox/60.0"
```

**Comment by Oleszkiewicz** [comment]
note 1) I don't have anything in /log, this stack trace was taken from the console after starting the containers with docker-compose up

note 2) Actually there is a way to dodge this bug - after entering a non existing domain name for the exchange connection I get the opportunity to input the Exchange connection endpoint, after manually putting the endpoint there - everything works. (While autodiscover for this domain is properly configured and works ok)

**Comment by thorsteneckel** [comment]
Thanks! Yes, this is a bug and should get properly handled by Zammad. I don't know why the used gem for autodiscovery does not work here. But it should be always possible to enter the endpoint manually like you did as described in `note 2)`.

Therefore we need to rescue from `Errno::EADDRNOTAVAIL` and return `nil` in that case.

Could you please send us the URL you've used to support@zammad.com so we can test it? Just refer to this issue and me. Thanks!

**Comment by rlue** [comment]
@Oleszkiewicz: as @thorsteneckel pointed out, we're using [a third-party gem to interface with Microsoft's autodiscovery service](https://github.com/WinRb/autodiscover). It looks like this problem originates in this gem. To confirm this, try running the following command in your Zammad container:

```
# should output the URL of your Exchange server endpoint
$ zammad run rails r 'p Autodiscover::Client.new(email: "<email address>", password: "<password>").autodiscover.ews_url'
```

If the line above actually works and successfully outputs your EWS endpoint, then there's truly something wrong with Zammad! But if not (which is the more likely scenario), you may want to [check the issues on that repo](https://github.com/WinRb/autodiscover/issues) for any that resemble your own and add a comment sharing your experience. The maintainer looks pretty unresponsive, though. :confused:

We're currently working on improvements to the UI to ensure that you don't have to rely on any funny tricks to "dodge" this bug in the future, but until the underlying bug is fixed upstream, the best we can do is allow you to enter the endpoint URL manually.

Please let us know what you find out! (FWIW, we're happy to debug the upstream issue with you over at that repo too.)

**Comment by rlue** [comment]
> Actually there is a way to dodge this bug - after entering a non existing domain name for the exchange connection I get the opportunity to input the Exchange connection endpoint, after manually putting the endpoint there - everything works. (While autodiscover for this domain is properly configured and works ok)

This problem has been fixed by 2e4b712. (_I.e.,_ autodiscovery will still fail, but you'll be prompted to enter the endpoint URL directly now instead of having to resort to this "trick".)

We're still happy to debug the autodiscovery failure with you, but that'll have to happen on the [autodiscover gem's project issue tracker](https://github.com/WinRb/autodiscover/issues).
